[pre-6.18] realtek: prom.c initialization fixes.

plappermaul opened the pull request:
This series fixes several shortcomings in prom.c. Especially boot stalls on RLT930x devices with kernel 6.18 and more than 256 MB memory.

plappermaul committed:
realtek: harmonize variables in prom.c

There is mix of variable naming in prom.c. Use a fixed
prefix of "rtl_" for all static variables. Additionally
remove the prefix from rtl83xx_set_system_type as it is
a generic function for all SoC types.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/21327
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

plappermaul committed:
realtek: determine memory size during initialization

For proper highmem initialization on RTL930x the size of the
installed memory is needed during early bootup. Enhance the
soc_info structure and fill the data from the registers.

While we are here remove the obsolete compatible variable from
the soc_info structure.

Adapt boot message to show the memory size.

old: SoC Type: Realtek RTL9301 rev B (6487)
new: Realtek RTL9301 rev B (6487) SoC with 512 MB

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/21327
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

plappermaul committed:
realtek: RTL930x: setup highmem registers during init

For some reason the highmem configuration of RTL930x
devices was totally missed until now. Take over the
setup from the SDK. This will avoid boot stalls on
switches with more than 256 MB RAM when switching
over to kernel 6.18.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/21327
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

jameywine commented on target/linux/realtek/files-6.12/arch/mips/rtl838x/prom.c:
```diff
@@ -223,19 +240,71 @@ static void __init rtl83xx_set_system_type(void)
+	} else {
+		dcr = soc_r32(RTL83XX_DRAM_CONFIG_REG);
+		bits = ((dcr >> 28) & 0x3) + ((dcr >> 24) & 0x3) +
+		       ((dcr >> 20) & 0xf) + ((dcr >> 16) & 0xf) + 20;
```
While i don't know much about the RTL83XX, this looks very much the same as RTL960X dram detection and also located at 0x1004 the same way, except for one more detail:
```c
    total_bit = 0;
    total_bit += ((dcr>>24)&0x3); //bus width
    total_bit += ((dcr>>20)&0x7)+11; //row count
    total_bit += ((dcr>>16)&0x7)+8 ; //col count
    total_bit += ((dcr>>28)&0x3)+1;  //bank count
    total_bit += (dcr>>15)&1;        //Dram Chip Select

    return ((1<<total_bit));
```
from 
https://github.com/jameywine/GPL-for-GP3000/blob/main/linux-5.10.x/arch/mips/rtl9607c/prom.c#L111

There is addition of dram chip select bit. Would it hurt to include it to the bits calculation here or is it ultimately unnecessary?

plappermaul committed:
realtek: setup cpu port in soc_info structure

A lot of soc_info usage has been reorganized. Nevertheless there
are some consumers left. A very critical one is the dsa/qos coding.
It makes use of the cpu_port in this shared structure. This is
totally broken as that info is never properly initialized. Fill
the cpu_port according to the identified hardware.

Remark: Looking at the prom.c history soc_info.cpu_port was never
setup since the beginning of time. So no "fixes" tag here.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/21327
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

plappermaul commented:
> But maybe nbd168's commits regarding setting a fitting offset to allow devices to boot without highmem might still be useful.

Only trying to fix things without performance in mind. Feel free to start exploration of alternative memory mapping offsets from here.

Neustradamus commented:
@plappermaul: Nice, thanks!

jonasjelonek reviewed (approved):
works with 6.18 on my SKS8310-8X with 512MB.
